subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c

```diff
@@ -0,0 +1,147 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <string.h>
+#include <errno.h>
+#include <cache.h>
+#include <ipc/ipc_icmsg_buf.h>
+
+
+/* Helpers */
+static uint32_t idx_occupied(uint32_t len, uint32_t a, uint32_t b)
+{
+	/* It is implicitly assumed a and b cannot differ by more then len. */
+	return (b > a) ? (len - (b - a)) : (a - b);
+}
+
+static uint32_t idx_cut(uint32_t len, uint32_t idx)
+{
+	/* It is implicitly assumed a and b cannot differ by more then len. */
+	return (idx >= len) ? (idx - len) : (idx);
+}
+
+int icmsg_buf_write(struct icmsg_buf *ib, const char *buf, uint8_t len)
+{
+	/* The length of buffer is immutable - avoid reloading. */
+	const uint32_t iblen = ib->len;
+
+	/* rx_idx == wr_idx means the buffer is empty.
+	 * Max bytes that can be stored is so len - 1.
+	 * Additionally sizeof(len) is reserved for storing len.
+	 */
+	const uint32_t max_len = iblen - 1 - sizeof(len);
+
+	sys_cache_data_range(ib, sizeof(*ib), K_CACHE_INVD);
+	__sync_synchronize();
+
+	uint32_t wr_idx = ib->wr_idx;
+	uint32_t rd_idx = ib->rd_idx;
+
+	if ((len > max_len) || (len == 0)) {
+		/* Incorrect call. */
+		return -EINVAL;
+	}
+
+	uint32_t avail = max_len - idx_occupied(iblen, wr_idx, rd_idx);
+
+	if (len > avail) {
+		/* No free space. */
+		return -ENOMEM;
+	}
+
+	/* Store info about the message length. */
+	ib->data[wr_idx] = len;
+	wr_idx = idx_cut(iblen, wr_idx + sizeof(len));
+
+	/* Write until the end of the buffer. */
+	uint32_t sz = MIN(len, iblen - wr_idx);
+
+	memcpy(&ib->data[wr_idx], buf, sz);
+	__sync_synchronize();
+	sys_cache_data_range(&ib->data[wr_idx], sz, K_CACHE_WB);
+
+	if (len > sz) {
+		/* Write remaining data at the buffer head. */
+		memcpy(&ib->data[0], buf + sz, len - sz);
+		__sync_synchronize();
+		sys_cache_data_range(&ib->data[0], len - sz, K_CACHE_WB);
+	}
+
+	/* Update write index - make other side aware data was written. */
+	wr_idx = idx_cut(iblen, wr_idx + len);
+	ib->wr_idx = wr_idx;
+
+	__sync_synchronize();
+	sys_cache_data_range(ib, sizeof(*ib), K_CACHE_WB);
+
+	return len;
+}
+
+int icmsg_buf_read(struct icmsg_buf *ib, char *buf, uint8_t len)
+{
+	/* The length of buffer is immutable - avoid reloading. */
+	const uint32_t iblen = ib->len;
+
+	sys_cache_data_range(ib, sizeof(*ib), K_CACHE_INVD);
+	__sync_synchronize();
+
+	uint32_t rd_idx = ib->rd_idx;
+	uint32_t wr_idx = ib->wr_idx;
+
+	if (rd_idx == wr_idx) {
+		/* The buffer is empty. */
+		return 0;
+	}
+
+	sys_cache_data_range(&ib->data[rd_idx], sizeof(ib->data[rd_idx]), K_CACHE_INVD);
+	__sync_synchronize();
+
+	uint8_t mlen = ib->data[rd_idx];
+
+	if (len < mlen) {
+		if (len == 0) {
+			return mlen;
+		} else {
+			return -EINVAL;
+		}
+	}
+
+	if (!buf) {
+		return -EINVAL;
+	}
+
+	uint32_t bytes_stored = idx_occupied(iblen, wr_idx, rd_idx);
+
+	if (bytes_stored < mlen + sizeof(mlen)) {
+		/* Part of message not available. Should not happen. */
+		__ASSERT_NO_MSG(false);
```

Comment:
Try to convince all Zephyr product makers to disable assertions in their products ;) I'm missing separated `debug_assert!()` and `assert!()` in C.

Reply:
Assert is really not substitute for error checking - if somebody uses this in production builds they have bigger problems. I won't try to convince anybody how to make products (unless they pay me to do so ;)).

Zephyr is wise enough to emit a noticeable warning message during build time that asserts are enabled.

We could remove this assert but that will make it harder to discover bugs - something in the higher layer (in the sample) would have to deal with the potential issue.
Anyway, if you think this is needed to get an approval I will not object :)

Reply:
My point was to raise awareness of this potential attack. I'm OK to merge it with it, because as you described asserts are disabled in production builds, and even if enabled DoS is not catastrophic.